Added AppendStoredProcedureCall

Camel-RD opened the pull request:
Added AppendStoredProcedureCall override to support InsertUsingStoredProcedure, UpdateUsingStoredProcedure, DeleteUsingStoredProcedure in model builder.

cincuranet commented:
Sorry for late reply, I was busy with work. Can you maybe add some simple tests for calling insert/update/delete SPs, just to have basic validation that the SQL is correct. Something similar to what is in [E2E tests](https://github.com/FirebirdSQL/NETProvider/tree/master/src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEnd).

Camel-RD commented:
It looks like EXECUTE PROCEDURE SP_ABC p1, p2 will not work properly with EF concurrency checks. Will have to change it to SELECT * FROM SP_ABC(p1, p2, p3).

Camel-RD commented:
AppendStoredProcedureCall fixed and tests added

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/DeleteTestsUsingSP.cs:
```diff
@@ -0,0 +1,150 @@
+			insertEntityConf.Property(x => x.Name).HasColumnName("NAME");
+			insertEntityConf.ToTable("TEST_DELETE_USP");
+			modelBuilder.Entity<DeleteEntity>().DeleteUsingStoredProcedure("SP_TEST_DELETE",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasRowsAffectedResultColumn();
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/DeleteTestsUsingSP.cs:
```diff
@@ -0,0 +1,150 @@
+				.IsRowVersion();
+			insertEntityConf.ToTable("TEST_DELETE_CONCURRENCY_USP");
+			modelBuilder.Entity<ConcurrencyDeleteEntity>().DeleteUsingStoredProcedure("SP_TEST_DELETE_CONCURRENCY",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Stamp);
+				storedProcedureBuilder.HasRowsAffectedResultColumn();
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/DeleteTestsUsingSP.cs:
```diff
@@ -0,0 +1,150 @@
+					 begin
+					     delete from test_delete_concurrency_usp where id = :pid and stamp = :pstamp;
+					     rowcount = row_count;
+					     if (rowcount > 0) then suspend;
```
Put `suspend` on new line.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/InsertTestsUsingSP.cs:
```diff
@@ -0,0 +1,321 @@
+			insertEntityConf.Property(x => x.Name).HasColumnName("NAME");
+			insertEntityConf.ToTable("TEST_INSERT_USP");
+			modelBuilder.Entity<InsertEntity>().InsertUsingStoredProcedure("SP_TEST_INSERT",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasParameter(x => x.Id);
+				storedProcedureBuilder.HasParameter(x => x.Name);
+				storedProcedureBuilder.HasResultColumn(x => x.Id);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/InsertTestsUsingSP.cs:
```diff
@@ -0,0 +1,321 @@
+			insertEntityConf.Property(x => x.Name).HasColumnName("NAME");
+			insertEntityConf.ToTable("TEST_INSERT_IDENTITY_USP");
+			modelBuilder.Entity<IdentityInsertEntity>().InsertUsingStoredProcedure("SP_TEST_INSERT_IDENTITY",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasParameter(x => x.Name);
+				storedProcedureBuilder.HasResultColumn(x => x.Id);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/InsertTestsUsingSP.cs:
```diff
@@ -0,0 +1,321 @@
+			insertEntityConf.Property(x => x.Name).HasColumnName("NAME");
+			insertEntityConf.ToTable("TEST_INSERT_SEQUENCE_USP");
+			modelBuilder.Entity<SequenceInsertEntity>().InsertUsingStoredProcedure("SP_TEST_INSERT_SEQUENCE",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasParameter(x => x.Name);
+				storedProcedureBuilder.HasResultColumn(x => x.Id);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/InsertTestsUsingSP.cs:
```diff
@@ -0,0 +1,321 @@
+				.ValueGeneratedOnAdd();
+			insertEntityConf.ToTable("TEST_INSERT_DEVAULTVALUES_USP");
+			modelBuilder.Entity<DefaultValuesInsertEntity>().InsertUsingStoredProcedure("SP_TEST_INSERT_DEFAULTVALUES",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasResultColumn(x => x.Id);
+				storedProcedureBuilder.HasResultColumn(x => x.Name);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/InsertTestsUsingSP.cs:
```diff
@@ -0,0 +1,321 @@
+				.ValueGeneratedOnAddOrUpdate();
+			insertEntityConf.ToTable("TEST_INSERT_2COMPUTED_USP");
+			modelBuilder.Entity<TwoComputedInsertEntity>().InsertUsingStoredProcedure("SP_TEST_INSERT_2COMPUTED",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasParameter(x => x.Name);
+				storedProcedureBuilder.HasResultColumn(x => x.Id);
+				storedProcedureBuilder.HasResultColumn(x => x.Computed1);
+				storedProcedureBuilder.HasResultColumn(x => x.Computed2);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+			insertEntityConf.Property(x => x.Bar).HasColumnName("BAR");
+			insertEntityConf.ToTable("TEST_UPDATE_USP");
+			modelBuilder.Entity<UpdateEntity>().UpdateUsingStoredProcedure("SP_TEST_UPDATE",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasParameter(x => x.Foo);
+				storedProcedureBuilder.HasParameter(x => x.Bar);
+				storedProcedureBuilder.HasRowsAffectedResultColumn();
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+				.FirstAsync();
+			Assert.AreEqual("test", value.Foo);
+			Assert.AreEqual("test", value.Bar);
+			//Assert.AreNotEqual("test", value.Bar); //this has to fail
```
Huh?

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+				.ValueGeneratedOnAddOrUpdate();
+			insertEntityConf.ToTable("TEST_UPDATE_COMPUTED_USP");
+			modelBuilder.Entity<ComputedUpdateEntity>().UpdateUsingStoredProcedure("SP_TEST_UPDATE_COMPUTED",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasParameter(x => x.Foo);
+				storedProcedureBuilder.HasParameter(x => x.Bar);
+				storedProcedureBuilder.HasResultColumn(x => x.Computed);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+					     update test_update_computed_usp set foo = :pfoo, bar = :pbar
+					     where id = :pid
+					     returning computed into :rcomputed;
+					     if (row_count > 0) then suspend;
```
Put `suspend` on new line.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+				.IsRowVersion();
+			insertEntityConf.ToTable("TEST_UPDATE_CONCURRENCY_USP");
+			modelBuilder.Entity<ConcurrencyUpdateEntity>().UpdateUsingStoredProcedure("SP_TEST_UPDATE_CONCURRENCY",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasParameter(x => x.Foo);
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Stamp);
+				storedProcedureBuilder.HasResultColumn(x => x.Stamp);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+				.IsConcurrencyToken();
+			insertEntityConf.ToTable("TEST_UPDATE_CONCURRENCY_NG_USP");
+			modelBuilder.Entity<ConcurrencyUpdateNoGeneratedEntity>().UpdateUsingStoredProcedure("SP_TEST_UPDATE_CONCURRENCY_NG",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasParameter(x => x.Foo);
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Stamp);
+				storedProcedureBuilder.HasRowsAffectedResultColumn();
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+					     update test_update_concurrency_ng_usp set foo = :pfoo, stamp = current_timestamp
+					     where id = :pid and stamp = :pstamp;
+					     rowcount = row_count;
+					     if (rowcount > 0) then suspend;
```
Put `suspend` on new line.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+				.ValueGeneratedOnAddOrUpdate();
+			insertEntityConf.ToTable("TEST_UPDATE_2COMPUTED_USP");
+			modelBuilder.Entity<TwoComputedUpdateEntity>().UpdateUsingStoredProcedure("SP_TEST_UPDATE_2COMPUTED",
+			storedProcedureBuilder =>
+			{
+				storedProcedureBuilder.HasOriginalValueParameter(x => x.Id);
+				storedProcedureBuilder.HasParameter(x => x.Foo);
+				storedProcedureBuilder.HasParameter(x => x.Bar);
+				storedProcedureBuilder.HasResultColumn(x => x.Computed1);
+				storedProcedureBuilder.HasResultColumn(x => x.Computed2);
+			});
```
Indent.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEndUsingSP/UpdateTestsUsingSP.cs:
```diff
@@ -0,0 +1,352 @@
+					     update test_update_2computed_usp set foo = :pfoo, bar = :pbar
+					     where id = :pid
+					     returning computed1, computed2 into :rcomputed1, :rcomputed2;
+					     if (row_count > 0) then suspend;
```
Put `suspend` on new line.

cincuranet commented on src/FirebirdSql.EntityFrameworkCore.Firebird/Update/Internal/FbUpdateSqlGenerator.cs:
```diff
@@ -252,4 +253,71 @@ void AppendSqlLiteral(StringBuilder commandStringBuilder, object value, IPropert
+			}
+		}
+
+		commandStringBuilder.Append("SELECT * FROM ");
```
I would prefer here to be explicit about column names.

cincuranet reviewed (changes requested):
Some minor pieces, otherwise LGTM in general.

Camel-RD commented:
Explicit column names means that we cannot freely choose the results column names in SPs, but we can pick names for input parameters. Also, EFC uses result values by their positions and ignores any given result column names.

cincuranet commented:
Not sure why you added commit 53fccb6 , but that looks unrelated and needs to be removed.

cincuranet commented:
> Explicit column names means that we cannot freely choose the results column names in SPs, but we can pick names for input parameters. Also, EFC uses result values by their positions and ignores any given result column names.

I don't follow. Can you elaborate?

Camel-RD commented:
> > Explicit column names means that we cannot freely choose the results column names in SPs, but we can pick names for input parameters. Also, EFC uses result values by their positions and ignores any given result column names.
> 
> I don't follow. Can you elaborate?

Not to use explicit names is how it's done in SQL Server implementation.

If explicit column names are used, there would also be an issue with naming the column for the rows affected value. The name could be ROWSCOUNT, but it should be properly document so developers writing stored procedures can easily find this information.

cincuranet commented:
> Not to use explicit names is how it's done in SQL Server implementation.

SQL Server implementation uses `EXEC` and SQL Server does not have selectable stored procedures. 

> If explicit column names are used, there would also be an issue with naming the column for the rows affected value. The name could be ROWSCOUNT, but it should be properly documented so developers writing stored procedures can easily find this information.

Yes. But same can be said about the whole SP structure. It needs to fit into what EF expects.

Camel-RD commented:
Updated the code to use explicit column names.

And i saw some additional issues with this AppendStoredProcedureCall code.

When stored procedure doesn't have any return values defined SELECT * FROM SP() will produce an SQL error. In this case EXECUTE PROCEDURE should be used.

Also AppendStoredProcedureCall code should probably throw some exceptions when output parameters or return values are defined.

Made changes to fix these issues and added a test for case when isn't auto generetad and insert procedure doesn't return any values.


> SQL Server implementation uses `EXEC` and SQL Server does not have selectable stored procedures.
Hm, SQLServer stored procedures have 3 ways to return values including as selectable procedure.

> Yes. But same can be said about the whole SP structure. It needs to fit into what EF expects.
Even if EF Core documentation is lacking, its possible to gues what EF is expecting from exceptions it throws.
